File: day4/puzzle1.cpp
#include <bits/stdc++.h>
using namespace std;

vector<string> read_input(){
    ifstream fin("input.txt");
    vector<string> res;
    string s;
    while (getline(fin, s)){
        res.push_back(s);
    }
    return res;
}

int solve(const vector<string>& matrix){
    int n = matrix.size();
    int m = matrix[0].size();
    int res = 0;
    for(int i = 0; i < n; ++i){
        for(int j = 0; j < m; ++j){
            if (matrix[i][j] == 'X'){
                if (j + 3 < m && matrix[i][j + 1] == 'M' && matrix[i][j + 2] == 'A' && matrix[i][j + 3] == 'S'){
                    res++;
                }
                if (j - 3 >= 0 && matrix[i][j - 1] == 'M' && matrix[i][j - 2] == 'A' && matrix[i][j - 3] == 'S'){
                    res++;
                }
                if (i + 3 < n && matrix[i + 1][j] == 'M' && matrix[i + 2][j] == 'A' && matrix[i + 3][j] == 'S'){
                    res++;
                }
                if (i - 3 >= 0 && matrix[i - 1][j] == 'M' && matrix[i - 2][j] == 'A' && matrix[i - 3][j] == 'S'){
                    res++;
                }
                if (i + 3 < n && j + 3 < m && matrix[i + 1][j + 1] == 'M' && matrix[i + 2][j + 2] == 'A' && matrix[i + 3][j + 3] == 'S'){
                    res++;
                }
                if (i - 3 >= 0 && j - 3 >= 0 && matrix[i - 1][j - 1] == 'M' && matrix[i - 2][j - 2] == 'A' && matrix[i - 3][j - 3] == 'S'){
                    res++;
                }
                if (i + 3 < n && j - 3 >= 0 && matrix[i + 1][j - 1] == 'M' && matrix[i + 2][j - 2] == 'A' && matrix[i + 3][j - 3] == 'S'){
                    res++;
                }
                if (i - 3 >= 0 && j + 3 < m && matrix[i - 1][j + 1] == 'M' && matrix[i - 2][j + 2] == 'A' && matrix[i - 3][j + 3] == 'S'){
                    res++;
                }
            }
        }
    }
    return res;
}

int main(){
    vector<string> matrix = read_input();
    cout << solve(matrix) << endl;
    return 0;
}

File: day13/puzzle1.py
from sympy import symbols, Eq
import sympy
import regex as re
def read_input(str):
    with open(str, "r") as f:
        pattern = r"""
            Button\sA:\sX\+(\d+),\sY\+(\d+)\n
            Button\sB:\sX\+(\d+),\sY\+(\d+)\n
            Prize:\sX=(\d+),\sY=(\d+)
        """
        lines = f.read()
        matches = re.findall(pattern, lines, re.VERBOSE)
        results = [
        {
            "a": int(match[0]),
            "b": int(match[2]),
            "c": int(match[4]),
            "d": int(match[1]),
            "e": int(match[3]),
            "f": int(match[5])
        }
        for match in matches
    ]

    return results

def solve(equations):
    res = 0
    for eq in equations:
        x, y = symbols("x y")
        eq1 = Eq(eq["a"] * x + eq["b"] * y, eq["c"])
        eq2 = Eq(eq["d"] * x + eq["e"] * y, eq["f"])
        sol = sympy.solve((eq1, eq2), (x, y))
        if sol[x] < 0 or sol[y] < 0:
            continue
        if sol[x] % 1 != 0 or sol[y] % 1 != 0:
            continue
        res += 3 * sol[x] + sol[y]
    return res

if __name__ == "__main__":
    equations = read_input("input.txt")
    print(solve(equations))